src/evg_config.rs

```diff
@@ -0,0 +1,848 @@
+use anyhow::{bail, Result};
+use lazy_static::lazy_static;
+use regex::Regex;
+use shrub_rs::models::commands::FunctionCall;
+use shrub_rs::models::params::ParamValue;
+use std::{collections::HashMap, path::Path, process::Command};
+
+use shrub_rs::models::commands::EvgCommand::Function;
+use shrub_rs::models::{project::EvgProject, task::EvgTask, variant::BuildVariant};
+
+use crate::evergreen_names::{GENERATE_RESMOKE_TASKS, IS_FUZZER};
+use crate::task_name::remove_gen_suffix;
+
+const REQUIRED_PREFIX: &str = "-required";
+
+lazy_static! {
+    /// Regular expression for finding expansions.
+    ///   `${expansion}` or `${expansion|default_value}`
+    static ref EXPANSION_RE: Regex =
+        Regex::new(r"\$\{(?P<id>[a-zA-Z0-9_]+)(\|(?P<default>.*))?}").unwrap();
+}
+
+pub trait EvgConfigService {
+    /// Get a map of build variant names to build variant definitions.
+    fn get_build_variant_map(&self) -> HashMap<String, &BuildVariant>;
+
+    /// Get a map of task name to task definitions.
+    fn get_task_def_map(&self) -> HashMap<String, &EvgTask>;
+
+    /// Get a list of build variants with the required build variants at the start.
+    fn sort_build_variants_by_required(&self) -> Vec<String>;
+}
+
+/// Items needed to implement an evergreen configuration service.
+pub struct EvgProjectConfig {
+    /// Shrub representation of the evg project.
+    evg_project: EvgProject,
+}
+
+impl EvgProjectConfig {
+    /// Create a new instance of an EvgConfigService.
+    ///
+    /// # Parameters
+    ///
+    /// * `evg_project_location` - Path to evergreen project configuration to load.
+    pub fn new(evg_project_location: &Path) -> Result<Self> {
+        let evg_project = get_project_config(evg_project_location)?;
+        Ok(Self { evg_project })
+    }
+}
+
+impl EvgConfigService for EvgProjectConfig {
+    /// Get a map of build variant names to build variant definitions.
+    fn get_build_variant_map(&self) -> HashMap<String, &BuildVariant> {
+        self.evg_project.build_variant_map()
+    }
+
+    /// Get a map of task name to task definitions.
+    fn get_task_def_map(&self) -> HashMap<String, &EvgTask> {
+        self.evg_project.task_def_map()
+    }
+
+    /// Get a list of build variants with the required build variants at the start.
+    fn sort_build_variants_by_required(&self) -> Vec<String> {
+        let build_variant_map = self.get_build_variant_map();
+        let mut build_variants: Vec<String> = build_variant_map
+            .keys()
+            .into_iter()
+            .filter_map(|bv| {
+                if bv.ends_with(REQUIRED_PREFIX) {
+                    Some(bv.to_string())
+                } else {
+                    None
+                }
+            })
+            .collect();
+
+        build_variants.extend::<Vec<String>>(
```

Comment:
The compiler can't tell we are trying to extend with a `Vec<String>` (due to using `collect` which support many types), so we have to tell it here.